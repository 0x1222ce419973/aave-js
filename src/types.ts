import { TransactionObject } from 'web3/eth/types'

export type ResponseCodes = 200 | 401 | 404 | 500 | 504

type Currency = {
  value: string
  description: string
  subChoices?: Currency[]
}

export interface LoanMetadata {
  mediums: Currency[]
  collaterals: Currency[]
  durationRange: { min: number; max: number }
}

export interface BaseLoanModel {
  moe: string
  loanAmount: number
  collateralType: string
  collateralAmount: number
  mpr: number
  duration: number
}

export interface LoanRequestModel extends BaseLoanModel {
  loanAddress?: string
  state?: string
  borrower?: string
  fundedAmount?: number
  outstandingLoanAmount: number
  startTimeoutTime?: number
  isPeggedLoan: boolean
  isCrowdLendingLoan: boolean
  peggedCurrency?: string
  peggedMedium?: string
  type: string
}

export interface BaseResponse {
  data?: any
  error?: string
  code: ResponseCodes
}

export interface LoanTransactionResponse extends BaseResponse {
  data?: TransactionObject<any>
}

export interface LoanRequestResponse extends BaseResponse {
  data?: LoanRequestModel
}

export interface LoanMetadataResponse extends BaseResponse {
  data?: LoanMetadata
}

export interface LoansAddressesResponse extends BaseResponse {
  data?: string[]
}

export interface LoanAddressesByBorrowerResponse extends BaseResponse {
  data?: {
    requestsAsBorrower: string[]
    requestsAsLender: string[]
  }
}

export interface LoanAPIInstance {
  create(creatorWalletAddress: string, params: BaseLoanModel): Promise<LoanTransactionResponse>
  placeCollateral(loanAddress: string, borrowerAddress: string): Promise<LoanTransactionResponse>
  fund(loanAddress: string, lenderAddress: string, amount: number): Promise<LoanTransactionResponse>
  payback(loanAddress: string, borrowerAddress: string): Promise<LoanTransactionResponse>
  getLoanData(loanAddress: string): Promise<LoanRequestResponse>
  getAllAddresses(): Promise<LoansAddressesResponse>
  getLoansByBorrower(borrowerAddress: string): Promise<LoanAddressesByBorrowerResponse>
  getMetadata(): Promise<LoanMetadataResponse>
}

export interface MarketplaceInstance {
  request: LoanAPIInstance
}

declare const Marketplace: MarketplaceInstance

export default Marketplace
