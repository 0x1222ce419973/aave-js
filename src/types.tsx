export enum BorrowRateMode {
  None = 'None',
  Stable = 'Stable',
  Variable = 'Variable',
}

export type ReserveData = {
  id: string;
  name: string;
  symbol: string;
  decimals: number;
  isActive: boolean;
  usageAsCollateralEnabled: number;
  borrowingEnabled: number;
  stableBorrowRateEnabled: number;
  baseLTVasCollateral: string;
  liquidityIndex: string;
  reserveLiquidationThreshold: string;
  variableBorrowIndex: string;
  availableLiquidity: string;
  stableBorrowRate: string;
  liquidityRate: string;
  totalBorrows: string;
  totalBorrowsStable: string;
  totalBorrowsVariable: string;
  totalLiquidity: string;
  utilizationRate: string;
  variableBorrowRate: string;
  lastUpdateTimestamp: number;
  aToken: {
    id: string;
  };
  price: {
    priceInEth: string;
  };
};

export type UserReserveData = {
  principalATokenBalance: string;
  userBalanceIndex: string;
  redirectedBalance: string;
  interestRedirectionAddress: string;
  usageAsCollateralEnabledOnUser: string;
  borrowRate: string;
  borrowRateMode: BorrowRateMode;
  originationFee: string;
  principalBorrows: string;
  variableBorrowIndex: string;
  lastUpdateTimestamp: number;
  reserve: {
    id: string;
    name: string;
    symbol: string;
    decimals: number;
    liquidityRate: string;
    lastUpdateTimestamp: number;
    aToken: {
      id: string;
    };
  };
};

export type ComputedUserReserve = UserReserveData & {
  currentUnderlyingBalance: string;
  currentUnderlyingBalanceETH: string;
  currentUnderlyingBalanceUSD: string;

  currentBorrows: string;
  currentBorrowsETH: string;
  currentBorrowsUSD: string;

  principalBorrowsETH: string;
  principalBorrowsUSD: string;

  originationFeeETH: string;
  originationFeeUSD: string;
};

export type UserSummaryData = {
  id: string;
  totalLiquidityETH: string;
  totalCollateralETH: string;
  totalBorrowsETH: string;
  totalFeesETH: string;
  totalLiquidityUSD: string;
  totalCollateralUSD: string;
  totalBorrowsUSD: string;
  availableBorrowsETH: string;
  currentLoanToValue: string;
  currentLiquidationThreshold: string;
  maxAmountToWithdrawInEth: string;
  healthFactor: string;
  reservesData: ComputedUserReserve[];
};
