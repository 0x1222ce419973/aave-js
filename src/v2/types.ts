import BigNumber from 'bignumber.js';

export type ReserveRatesData = {
  id: string;
  symbol: string;
  paramsHistory: {
    variableBorrowIndex: string;
    liquidityIndex: string;
    timestamp: number;
  }[];
};

export type ReserveSupplyData = {
  totalScaledVariableDebt: string;
  variableBorrowIndex: string;
  variableBorrowRate: string;
  totalPrincipalStableDebt: string;
  averageStableRate: string;
  availableLiquidity: string;
  stableDebtLastUpdateTimestamp: number;
  lastUpdateTimestamp: number;
};

export type RewardsInformation = {
  rewardTokenAddress: string;
  rewardTokenDecimals: number;
  incentivePrecision: number;
  rewardTokenPriceEth: string;
  emissionEndTimestamp: number;
};

export type ReserveData = {
  id: string;
  underlyingAsset: string;
  name: string;
  symbol: string;
  decimals: number;
  isActive: boolean;
  isFrozen: boolean;
  usageAsCollateralEnabled: boolean;
  aTokenAddress: string;
  stableDebtTokenAddress: string;
  variableDebtTokenAddress: string;
  borrowingEnabled: boolean;
  stableBorrowRateEnabled: boolean;
  reserveFactor: string;
  baseLTVasCollateral: string;
  optimalUtilisationRate: string;
  stableRateSlope1: string;
  stableRateSlope2: string;
  averageStableRate: string;
  stableDebtLastUpdateTimestamp: number;
  baseVariableBorrowRate: string;
  variableRateSlope1: string;
  variableRateSlope2: string;
  liquidityIndex: string;
  reserveLiquidationThreshold: string;
  reserveLiquidationBonus: string;
  variableBorrowIndex: string;
  variableBorrowRate: string;
  avg30DaysVariableBorrowRate?: string;
  availableLiquidity: string;
  stableBorrowRate: string;
  liquidityRate: string;
  avg30DaysLiquidityRate?: string;
  totalPrincipalStableDebt: string;
  totalScaledVariableDebt: string;
  lastUpdateTimestamp: number;
  price: {
    priceInEth: string;
  };
};

export type IncentivizedReserve = {
  id: string;
  aEmissionPerSecond: string;
  vEmissionPerSecond: string;
  sEmissionPerSecond: string;
  totalLiquidity: string;
  totalVariableDebt: string;
  totalStableDebt: string;
  price: {
    priceInEth: string;
  };
};

export type ReserveIncentivesAPY = {
  id: string;
  aIncentivesAPY: string;
  vIncentivesAPY: string;
  sIncentivesAPY: string;
};

export type IncentivizedUserReserve = {
  aTokenincentivesUserIndex: string;
  vTokenincentivesUserIndex: string;
  sTokenincentivesUserIndex: string;
  reserve: {
    id: string;
  };
};

export type IncentivizedUserReserveData = {
  id: string;
  averageStableRate: string;
  stableDebtLastUpdateTimestamp: number;
  variableBorrowIndex: string;
  variableBorrowRate: string;
  avg30DaysVariableBorrowRate?: string;
  availableLiquidity: string;
  totalPrincipalStableDebt: string;
  totalScaledVariableDebt: string;
  lastUpdateTimestamp: number;
  aEmissionPerSecond: string;
  vEmissionPerSecond: string;
  sEmissionPerSecond: string;
  aIncentivesLastUpdateTimestamp: number;
  vIncentivesLastUpdateTimestamp: number;
  sIncentivesLastUpdateTimestamp: number;
  aTokenIncentivesIndex: string;
  vTokenIncentivesIndex: string;
  sTokenIncentivesIndex: string;

  underlyingBalance: string;
  variableBorrows: string;
  stableBorrows: string;
};

export type ComputedReserveData = {
  utilizationRate: string;
  totalStableDebt: string;
  totalVariableDebt: string;
  totalDebt: string;
  totalLiquidity: string;
} & ReserveData;

export type Supplies = {
  totalVariableDebt: BigNumber;
  totalStableDebt: BigNumber;
  totalLiquidity: BigNumber;
};

export type UserReserveData = {
  scaledATokenBalance: string;
  usageAsCollateralEnabledOnUser: boolean;
  scaledVariableDebt: string;
  variableBorrowIndex: string;
  stableBorrowRate: string;
  principalStableDebt: string;
  stableBorrowLastUpdateTimestamp: number;
  reserve: {
    id: string;
    underlyingAsset: string;
    name: string;
    symbol: string;
    decimals: number;
    liquidityRate: string;
    reserveLiquidationBonus: string;
    lastUpdateTimestamp: number;
  };
};

export type ComputedUserReserve = UserReserveData & {
  underlyingBalance: string;
  underlyingBalanceETH: string;
  underlyingBalanceUSD: string;

  variableBorrows: string;
  variableBorrowsETH: string;
  variableBorrowsUSD: string;

  stableBorrows: string;
  stableBorrowsETH: string;
  stableBorrowsUSD: string;

  totalBorrows: string;
  totalBorrowsETH: string;
  totalBorrowsUSD: string;
};

export type ComputedUserRewards = {
  reservesData: ReserveRewards[];
  totalRewards: string;
  totalRewardsETH: string;
  totalRewardsUSD: string;
};

export type ReserveRewards = {
  id: string;
  aTokenRewards: string;
  aTokenRewardsETH: string;
  aTokenRewardsUSD: string;
  vTokenRewards: string;
  vTokenRewardsETH: string;
  vTokenRewardsUSD: string;
  sTokenRewards: string;
  sTokenRewardsETH: string;
  sTokenRewardsUSD: string;
  totalRewards: string;
  totalRewardsETH: string;
  totalRewardsUSD: string;
};

export type UserSummaryData = {
  id: string;
  totalLiquidityETH: string;
  totalLiquidityUSD: string;
  totalCollateralETH: string;
  totalCollateralUSD: string;
  totalBorrowsETH: string;
  totalBorrowsUSD: string;
  availableBorrowsETH: string;
  currentLoanToValue: string;
  currentLiquidationThreshold: string;
  healthFactor: string;
  reservesData: ComputedUserReserve[];
};
