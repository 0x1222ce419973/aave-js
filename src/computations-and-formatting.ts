import BigNumber from 'bignumber.js';

import {
  BigNumberValue,
  valueToBigNumber,
  valueToZDBigNumber,
} from './helpers/bignumber';
import {
  calculateAvailableBorrowsETH,
  calculateHealthFactorFromBalances,
  getCompoundedBalance,
  getCompoundedStableBalance,
  normalize,
  calculateAverageRate,
  LTV_PRECISION,
} from './helpers/pool-math';
import {
  ComputedUserReserve,
  ReserveData,
  UserReserveData,
  UserSummaryData,
  ReserveRatesData,
} from './types';

const ETH_DECIMALS = 18;
const USD_DECIMALS = 10;
const RAY_DECIMALS = 27;

function getEthAndUsdBalance(
  balance: BigNumberValue,
  priceInEth: BigNumberValue,
  decimals: number,
  usdPriceEth: BigNumberValue
): [string, string] {
  const balanceInEth = valueToZDBigNumber(balance)
    .multipliedBy(priceInEth)
    .dividedBy(10 ** decimals);
  const balanceInUsd = balanceInEth
    .multipliedBy(10 ** USD_DECIMALS)
    .dividedBy(usdPriceEth)
    .toFixed(0);
  return [balanceInEth.toString(), balanceInUsd];
}

function computeUserReserveData(
  poolReserve: ReserveData,
  userReserve: UserReserveData,
  usdPriceEth: BigNumberValue,
  currentTimestamp: number
): ComputedUserReserve {
  const {
    price: { priceInEth },
    decimals,
  } = poolReserve;
  const underlyingBalance = getCompoundedBalance(
    userReserve.principalATokenBalance,
    poolReserve.liquidityIndex,
    poolReserve.liquidityRate,
    poolReserve.lastUpdateTimestamp,
    currentTimestamp
  )
    .multipliedBy(
      valueToZDBigNumber(LTV_PRECISION).minus(poolReserve.reserveFactor)
    )
    .dividedBy(LTV_PRECISION)
    .toString();
  const [underlyingBalanceETH, underlyingBalanceUSD] = getEthAndUsdBalance(
    underlyingBalance,
    priceInEth,
    decimals,
    usdPriceEth
  );

  const variableBorrows = getCompoundedBalance(
    userReserve.principalVariableBorrows,
    poolReserve.variableBorrowIndex,
    poolReserve.variableBorrowRate,
    poolReserve.lastUpdateTimestamp,
    currentTimestamp
  ).toString();

  const [variableBorrowsETH, variableBorrowsUSD] = getEthAndUsdBalance(
    variableBorrows,
    priceInEth,
    decimals,
    usdPriceEth
  );

  const stableBorrows = getCompoundedStableBalance(
    userReserve.principalStableBorrows,
    userReserve.stableBorrowRate,
    userReserve.lastStableBorrowsUpdateTimestamp,
    currentTimestamp
  ).toString();

  const [stableBorrowsETH, stableBorrowsUSD] = getEthAndUsdBalance(
    stableBorrows,
    priceInEth,
    decimals,
    usdPriceEth
  );

  return {
    ...userReserve,
    underlyingBalance,
    underlyingBalanceETH,
    underlyingBalanceUSD,
    variableBorrows,
    variableBorrowsETH,
    variableBorrowsUSD,
    stableBorrows,
    stableBorrowsETH,
    stableBorrowsUSD,
    totalBorrows: valueToZDBigNumber(variableBorrows)
      .plus(stableBorrows)
      .toString(),
    totalBorrowsETH: valueToZDBigNumber(variableBorrowsETH)
      .plus(stableBorrowsETH)
      .toString(),
    totalBorrowsUSD: valueToZDBigNumber(variableBorrowsUSD)
      .plus(stableBorrowsUSD)
      .toString(),
  };
}

export function computeRawUserSummaryData(
  poolReservesData: ReserveData[],
  rawUserReserves: UserReserveData[],
  userId: string,
  usdPriceEth: BigNumberValue,
  currentTimestamp: number
): UserSummaryData {
  let totalLiquidityETH = valueToZDBigNumber('0');
  let totalCollateralETH = valueToZDBigNumber('0');
  let totalBorrowsETH = valueToZDBigNumber('0');
  let currentLtv = valueToBigNumber('0');
  let currentLiquidationThreshold = valueToBigNumber('0');

  const userReservesData = rawUserReserves
    .map(userReserve => {
      const poolReserve = poolReservesData.find(
        reserve => reserve.id === userReserve.reserve.id
      );
      if (!poolReserve) {
        throw new Error(
          'Reserve is not registered on platform, please contact support'
        );
      }
      const computedUserReserve = computeUserReserveData(
        poolReserve,
        userReserve,
        usdPriceEth,
        currentTimestamp
      );
      totalLiquidityETH = totalLiquidityETH.plus(
        computedUserReserve.underlyingBalanceETH
      );
      totalBorrowsETH = totalBorrowsETH
        .plus(computedUserReserve.variableBorrowsETH)
        .plus(computedUserReserve.stableBorrowsETH);

      // asset enabled as collateral
      if (
        poolReserve.usageAsCollateralEnabled &&
        userReserve.usageAsCollateralEnabledOnUser
      ) {
        totalCollateralETH = totalCollateralETH.plus(
          computedUserReserve.underlyingBalanceETH
        );
        currentLtv = currentLtv.plus(
          valueToBigNumber(
            computedUserReserve.underlyingBalanceETH
          ).multipliedBy(poolReserve.baseLTVasCollateral)
        );
        currentLiquidationThreshold = currentLiquidationThreshold.plus(
          valueToBigNumber(
            computedUserReserve.underlyingBalanceETH
          ).multipliedBy(poolReserve.reserveLiquidationThreshold)
        );
      }
      return computedUserReserve;
    })
    .sort((a, b) =>
      a.reserve.symbol > b.reserve.symbol
        ? 1
        : a.reserve.symbol < b.reserve.symbol
        ? -1
        : 0
    );

  if (currentLtv.gt(0)) {
    currentLtv = currentLtv
      .div(totalCollateralETH)
      .decimalPlaces(0, BigNumber.ROUND_DOWN);
  }
  if (currentLiquidationThreshold.gt(0)) {
    currentLiquidationThreshold = currentLiquidationThreshold
      .div(totalCollateralETH)
      .decimalPlaces(0, BigNumber.ROUND_DOWN);
  }

  const healthFactor = calculateHealthFactorFromBalances(
    totalCollateralETH,
    totalBorrowsETH,
    currentLiquidationThreshold
  );

  const totalCollateralUSD = totalCollateralETH
    .multipliedBy(10 ** USD_DECIMALS)
    .dividedBy(usdPriceEth)
    .toString();

  const totalLiquidityUSD = totalLiquidityETH
    .multipliedBy(10 ** USD_DECIMALS)
    .dividedBy(usdPriceEth)
    .toString();

  const totalBorrowsUSD = totalBorrowsETH
    .multipliedBy(10 ** USD_DECIMALS)
    .dividedBy(usdPriceEth)
    .toString();

  const availableBorrowsETH = calculateAvailableBorrowsETH(
    totalCollateralETH,
    totalBorrowsETH,
    currentLtv
  );

  return {
    totalLiquidityUSD,
    totalCollateralUSD,
    totalBorrowsUSD,
    id: userId,
    totalLiquidityETH: totalLiquidityETH.toString(),
    totalCollateralETH: totalCollateralETH.toString(),
    totalBorrowsETH: totalBorrowsETH.toString(),
    availableBorrowsETH: availableBorrowsETH.toString(),
    currentLoanToValue: currentLtv.toString(),
    currentLiquidationThreshold: currentLiquidationThreshold.toString(),
    healthFactor: healthFactor.toString(),
    reservesData: userReservesData,
  };
}

export function formatUserSummaryData(
  poolReservesData: ReserveData[],
  rawUserReserves: UserReserveData[],
  userId: string,
  usdPriceEth: BigNumberValue,
  currentTimestamp: number
): UserSummaryData {
  const userData = computeRawUserSummaryData(
    poolReservesData,
    rawUserReserves,
    userId,
    usdPriceEth,
    currentTimestamp
  );
  const userReservesData = userData.reservesData.map(
    ({ reserve, ...userReserve }): ComputedUserReserve => {
      const reserveDecimals = reserve.decimals;
      return {
        ...userReserve,
        reserve: {
          ...reserve,
          reserveLiquidationBonus: normalize(
            valueToBigNumber(reserve.reserveLiquidationBonus).minus(
              LTV_PRECISION
            ),
            4
          ),
          liquidityRate: normalize(reserve.liquidityRate, RAY_DECIMALS),
        },
        principalATokenBalance: normalize(
          userReserve.principalATokenBalance,
          reserveDecimals
        ),
        stableBorrowRate: normalize(userReserve.stableBorrowRate, RAY_DECIMALS),
        lastStableBorrowsUpdateTimestamp:
          userReserve.lastStableBorrowsUpdateTimestamp,
        variableBorrowIndex: normalize(
          userReserve.variableBorrowIndex,
          RAY_DECIMALS
        ),
        userBalanceIndex: normalize(userReserve.userBalanceIndex, RAY_DECIMALS),
        underlyingBalance: normalize(
          userReserve.underlyingBalance,
          reserveDecimals
        ),
        underlyingBalanceETH: normalize(
          userReserve.underlyingBalanceETH,
          ETH_DECIMALS
        ),
        underlyingBalanceUSD: normalize(
          userReserve.underlyingBalanceUSD,
          USD_DECIMALS
        ),
        stableBorrows: normalize(userReserve.stableBorrows, reserveDecimals),
        stableBorrowsETH: normalize(userReserve.stableBorrowsETH, ETH_DECIMALS),
        stableBorrowsUSD: normalize(userReserve.stableBorrowsUSD, USD_DECIMALS),
        variableBorrows: normalize(
          userReserve.variableBorrows,
          reserveDecimals
        ),
        variableBorrowsETH: normalize(
          userReserve.variableBorrowsETH,
          ETH_DECIMALS
        ),
        variableBorrowsUSD: normalize(
          userReserve.variableBorrowsUSD,
          USD_DECIMALS
        ),
        totalBorrows: normalize(userReserve.totalBorrows, reserveDecimals),
        totalBorrowsETH: normalize(userReserve.totalBorrowsETH, ETH_DECIMALS),
        totalBorrowsUSD: normalize(userReserve.totalBorrowsUSD, USD_DECIMALS),
      };
    }
  );
  return {
    id: userData.id,
    reservesData: userReservesData,
    totalLiquidityETH: normalize(userData.totalLiquidityETH, ETH_DECIMALS),
    totalLiquidityUSD: normalize(userData.totalLiquidityUSD, USD_DECIMALS),
    totalCollateralETH: normalize(userData.totalCollateralETH, ETH_DECIMALS),
    totalCollateralUSD: normalize(userData.totalCollateralUSD, USD_DECIMALS),
    totalBorrowsETH: normalize(userData.totalBorrowsETH, ETH_DECIMALS),
    totalBorrowsUSD: normalize(userData.totalBorrowsUSD, USD_DECIMALS),
    availableBorrowsETH: normalize(userData.availableBorrowsETH, ETH_DECIMALS),
    currentLoanToValue: normalize(userData.currentLoanToValue, 4),
    currentLiquidationThreshold: normalize(
      userData.currentLiquidationThreshold,
      4
    ),
    healthFactor: userData.healthFactor,
  };
}

export function formatReserves(
  reserves: ReserveData[],
  reserveIndexes30DaysAgo?: ReserveRatesData[]
): ReserveData[] {
  return reserves.map(reserve => {
    const reserve30DaysAgo = reserveIndexes30DaysAgo?.find(
      res => res.id === reserve.id
    )?.paramsHistory[0];

    return {
      ...reserve,
      price: {
        ...reserve.price,
        priceInEth: normalize(reserve.price.priceInEth, ETH_DECIMALS),
      },
      baseLTVasCollateral: normalize(
        reserve.baseLTVasCollateral,
        LTV_PRECISION
      ),
      reserveFactor: normalize(reserve.reserveFactor, LTV_PRECISION),
      variableBorrowRate: normalize(reserve.variableBorrowRate, RAY_DECIMALS),
      avg30DaysVariableBorrowRate: reserve30DaysAgo
        ? calculateAverageRate(
            reserve30DaysAgo.variableBorrowIndex,
            reserve.variableBorrowIndex,
            reserve30DaysAgo.timestamp,
            reserve.lastUpdateTimestamp
          )
        : undefined,
      avg30DaysLiquidityRate: reserve30DaysAgo
        ? calculateAverageRate(
            reserve30DaysAgo.liquidityIndex,
            reserve.liquidityIndex,
            reserve30DaysAgo.timestamp,
            reserve.lastUpdateTimestamp
          )
        : undefined,

      stableBorrowRate: normalize(reserve.stableBorrowRate, RAY_DECIMALS),
      liquidityRate: normalize(reserve.liquidityRate, RAY_DECIMALS),
      totalLiquidity: normalize(reserve.totalLiquidity, reserve.decimals),
      availableLiquidity: normalize(
        reserve.availableLiquidity,
        reserve.decimals
      ),
      liquidityIndex: normalize(reserve.liquidityIndex, RAY_DECIMALS),
      reserveLiquidationThreshold: normalize(
        reserve.reserveLiquidationThreshold,
        4
      ),
      reserveLiquidationBonus: normalize(
        valueToBigNumber(reserve.reserveLiquidationBonus).minus(LTV_PRECISION),
        4
      ),
      totalBorrows: normalize(reserve.totalBorrows, reserve.decimals),
      totalBorrowsVariable: normalize(
        reserve.totalBorrowsVariable,
        reserve.decimals
      ),
      totalBorrowsStable: normalize(
        reserve.totalBorrowsStable,
        reserve.decimals
      ),
      variableBorrowIndex: normalize(reserve.variableBorrowIndex, RAY_DECIMALS),
    };
  });
}
