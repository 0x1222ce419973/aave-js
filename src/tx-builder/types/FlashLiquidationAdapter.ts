import { tEthereumAddress } from '.';

export type Liquidation = {
  user: tEthereumAddress;
  collateralAsset: tEthereumAddress;
  borrowedAsset: tEthereumAddress;
  debtTokenCover: number;
  flashBorrowedAmount: number;
  onBehalfOf: tEthereumAddress;
  useEthPath: boolean;
};
