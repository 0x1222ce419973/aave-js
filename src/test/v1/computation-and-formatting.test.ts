import { ReserveData } from '../../v1/types';
import { formatReserves } from '../../v1/computations-and-formatting';

const mockReserve: ReserveData = {
  underlyingAsset: '0xff795577d9ac8bd7d90ee22b6c1703490b6512fd',
  name: '',
  symbol: 'DAI',
  decimals: 18,
  baseLTVasCollateral: '7500',
  reserveLiquidationThreshold: '8000',
  reserveLiquidationBonus: '10500',
  usageAsCollateralEnabled: true,
  borrowingEnabled: true,
  stableBorrowRateEnabled: true,
  isActive: true,
  liquidityIndex: '1000164447379610590574518134',
  variableBorrowIndex: '1000232854433711209646283880',
  liquidityRate: '26776200735312093055313462',
  variableBorrowRate: '38568743388028395681971229',
  stableBorrowRate: '109284371694014197840985614',
  lastUpdateTimestamp: 1606992400,
  // interestRateStrategyAddress: '0x1c4c4dD7F19738Fd7C21Fa7CbF9667710ff3Ba4c',
  availableLiquidity: '43133641118657852003256',
  // priceInEth: '1634050000000000',
  variableRateSlope1: '40000000000000000000000000',
  variableRateSlope2: '750000000000000000000000000',
  stableRateSlope1: '20000000000000000000000000',
  stableRateSlope2: '750000000000000000000000000',
  id:
    '0xff795577d9ac8bd7d90ee22b6c1703490b6512fd0x88757f2f99175387ab4c6a4b3067c77a695b0349',
  price: { priceInEth: '1634050000000000' },
  // fake data
  optimalUtilisationRate: '109284236984257451326752610',
  baseVariableBorrowRate: '109284236984257451326752610',
  totalBorrows: '0',
  totalBorrowsStable: '0',
  totalBorrowsVariable: '0',
  isFreezed: true,
  averageStableBorrowRate: '0',
  utilizationRate: '0',
  totalLiquidity: '0',
  aToken: {
    id: '0x0',
  },
};

describe('computations and formattings', () => {
  describe('formatReserves', () => {
    it('should return plausible results', () => {
      const formattedMockReserve = formatReserves([mockReserve])[0];
      expect(formattedMockReserve).toMatchSnapshot();
    });

    /**
     * Whenever we add a new asset there#s a chance that an asset has no paramsHistory from 30days ago
     * We should not throw if that's the case, but just ignore it
     */
    it("should not error When 30dago reserves doesn't contain paramsHistory", () => {
      formatReserves([mockReserve], [mockReserve as any]);
      formatReserves(
        [mockReserve],
        [{ ...mockReserve, paramsHistory: [] } as any]
      );
      // would be wrong if any of the above threw
      expect(true).toBe(true);
    });
  });
});
