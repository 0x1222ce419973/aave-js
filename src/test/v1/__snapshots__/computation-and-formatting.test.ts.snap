// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computations and formattings formatReserves should return plausible results 1`] = `
Object {
  "aToken": Object {
    "id": "0x0",
  },
  "availableLiquidity": "43133.641118657852003256",
  "averageStableBorrowRate": "0",
  "avg30DaysLiquidityRate": undefined,
  "avg30DaysVariableBorrowRate": undefined,
  "baseLTVasCollateral": "75",
  "baseVariableBorrowRate": "109284236984257451326752610",
  "borrowingEnabled": true,
  "decimals": 18,
  "id": "0xff795577d9ac8bd7d90ee22b6c1703490b6512fd0x88757f2f99175387ab4c6a4b3067c77a695b0349",
  "isActive": true,
  "isFreezed": true,
  "lastUpdateTimestamp": 1606992400,
  "liquidityIndex": "1.00016444737961059057",
  "liquidityRate": "0.02677620073531209306",
  "name": "",
  "optimalUtilisationRate": "109284236984257451326752610",
  "price": Object {
    "priceInEth": "0.00163405",
  },
  "reserveLiquidationBonus": "104",
  "reserveLiquidationThreshold": "80",
  "stableBorrowRate": "0.10928437169401419784",
  "stableBorrowRateEnabled": true,
  "stableRateSlope1": "20000000000000000000000000",
  "stableRateSlope2": "750000000000000000000000000",
  "symbol": "DAI",
  "totalBorrows": "0",
  "totalBorrowsStable": "0",
  "totalBorrowsVariable": "0",
  "totalLiquidity": "0",
  "underlyingAsset": "0xff795577d9ac8bd7d90ee22b6c1703490b6512fd",
  "usageAsCollateralEnabled": true,
  "utilizationRate": "0",
  "variableBorrowIndex": "1.00023285443371120965",
  "variableBorrowRate": "0.03856874338802839568",
  "variableRateSlope1": "40000000000000000000000000",
  "variableRateSlope2": "750000000000000000000000000",
}
`;
